src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+{
+    internal class PreloginPacketHandler : IHandler<PreLoginHandlerContext>
+    {
+        const int GUID_SIZE = 16;
+
+        const int PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES = 4;
+
+        const int PAYLOAD_LENGTH_SIZE_IN_BYTES = 2;
+
+        // EventSource counter
+        private static int s_objectTypeCount;
+
+        internal readonly int _objectID = Interlocked.Increment(ref s_objectTypeCount);
+
+        internal int ObjectID => _objectID;
+
+        public IHandler<PreLoginHandlerContext> NextHandler { get; set; }
+
+        public async ValueTask Handle(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            await PreloginPacketHandler.CreatePreLoginAndSend(context, isAsync, ct).ConfigureAwait(false);
+            if (context.HasError)
+            {
+                return;
+            }
+            await ReadPreLoginresponse(context, isAsync, ct).ConfigureAwait(false);
+
+            if (context.HasError)
+            {
+                return;
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Consume the Prelogin response from the server.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async Task ReadPreLoginresponse(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            context.ConnectionContext.MarsCapable = context.ConnectionContext.ConnectionString.MARS; // Assign default value
+            context.ConnectionContext.FedAuthRequired = false;
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+
+            // Look into the first byte to see if we are connecting to a 6.5 or earlier server.
+            int option = await tdsStream.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+
+            if (option == 0xaa)
+            {
+                // If the first byte is 0xAA, we are connecting to a 6.5 or earlier server, which
+                // is not supported.
+                throw SQL.InvalidSQLServerVersionUnknown();
+            }
+
+            byte[] payload = new byte[tdsStream.PacketDataLeft];
+
+            int payloadOffset = 0;
+            int offset = 0;
+
+            while (option != (byte)PreLoginOptions.LASTOPT)
+            {
+                switch (option)
+                {
+                    case (int)PreLoginOptions.VERSION:
+
+                        payloadOffset = payload[offset++] << 8 | payload[offset++];
+                        offset += PAYLOAD_LENGTH_SIZE_IN_BYTES; // Skip the payload length
+
+                        byte majorVersion = payload[payloadOffset];
+
+                        bool is2005OrLater = majorVersion >= 9;
```

Comment:
SqlClient's official support policy doesn't need to support older SQL Server versions. Specially with SqlClientX, our minimum SQL Server version can easily be SQL 2014 and above.
ref: [SQL version compatibility with Microsoft.Data.SqlClient](https://learn.microsoft.com/en-us/sql/connect/ado-net/sqlclient-driver-support-lifecycle?view=sql-server-ver16#sql-version-compatibility-with-microsoftdatasqlclient)

Reply:
@cheenamalhotra do you have any objections this being enforce this in the main branch? At present the .NET Framework version works with SQL 7.0+, and .NET Core works with SQL 2005+. I think it'd be best to introduce the enforcement before SqlClientX is exposed, to avoid a situation where SqlClientX is disabled in code to allow connections to an old database server, then never reenabled when the connection string is updated to point to a newer version.

I'm not sure 

SQL 2014 is a pretty low bar to reach, but 7.0 -> 2014 might be a large version jump for .NET Framework. I was thinking that SqlClient 6.0 could have a hard compatibility block on versions prior to 2005 (which immediately aligns .NET Core and Framework) and a soft block on 2005, 2008/R2 and 2012 (where out-of-support version compatibility can be enabled by a temporary AppContext switch.)

Alternatively, we could immediately instate a hard block on SQL Server versions prior to 2014 and say that connections to a legacy database server should be made via System.Data.SqlClient.

Reply:
@cheenamalhotra do you have any objections to this being enforced in the main branch? At present the .NET Framework version works with SQL 7.0+, and .NET Core works with SQL 2005+. I think it'd be best to introduce the enforcement before SqlClientX is exposed, to avoid a situation where SqlClientX is disabled in code to allow connections to an old database server, then never reenabled when the connection string is updated to point to a newer version.

SQL 2014 is a pretty low bar to reach, but I'm in two minds on how the transition to enforcement should be managed. On the one hand: SQL 7.0 -> 2014 is a large version jump for .NET Framework. Perhaps SqlClient 6.0 could have a hard compatibility block on versions prior to 2005 (which immediately aligns .NET Core and Framework) and a soft block on 2005, 2008/R2 and 2012 (where out-of-support version compatibility can be enabled by a temporary AppContext switch. This switch would probably be removed in SqlClient 7.0.)

On the other hand, System.Data.SqlClient is still around. We could say that connections to a legacy database server require the use of the legacy S.D.S library, and instate an immediate hard block on unsupported SQL Server versions on both versions of M.D.S. This would provide no way for .NET Core on Linux to connect to SQL Server 2005-2012 - but I'm not sure if that'd be a major concern.

Reply:
There's no reason that main needs to work on anything older than 2014. Scratch that. 2016, since 2014 EOLs July 9. Users who are running EOL servers can use old clients.

Reply:
Thanks - I'll work on the PR. Once that's in place, there'll also need to be a follow-up check for anything which references `Is2000`, `Is2005OrNewer`, `Is2008OrNewer`, SQLALTROW and SQLALTMETADATA and which might be removable.

Reply:
I removed these checks from Prelogin. An uber issue can be opened for the rest of the driver.